src/differentials.jl

```diff
@@ -272,6 +282,75 @@ function itself, when that function is not a closure.
 """
 const NO_FIELDS = DoesNotExist()
 
+
+"""
+    Composite{Primal, T} <: AbstractDifferential
+
+This type represents the differential for a `struct`/`NamedTuple`, or `Tuple`.
+`Primal` is the the corresponding primal type that this is a differential for.
+
+`Composite{Primal}` should have fields (technically properties), that match to a subset of the
+fields of the primal type; and each should be a differential type matching to the primal
+type of that field.
+Fields of the Primal that are not present in the Composite are treated as `Zero`.
+
+`T` is an implementation detail representing the backing data structure.
+For Tuple it will be a Tuple, and for everything else it will be a `NamedTuple`.
+It should not be passed in by user.
+"""
+struct Composite{Primal, T} <: AbstractDifferential
+    backing::T
+end
+# Note: If T is a Tuple, then Primal is also a Tuple (but potentially a different one, as it doesn't contain differentials)
+
+function Composite{Primal}(; kwargs...) where Primal
+    backing = (; kwargs...)  # construct as NamedTuple
+    return Composite{Primal, typeof(backing)}(backing)
+end
+
+function Composite{Primal}(args...) where Primal
+    return Composite{Primal, typeof(args)}(args)
+end
+
+function Base.show(io::IO, comp::Composite{Primal}) where Primal
+    print(io, "Composite{")
+    show(io, Primal)
+    print(io, "}")
+    # allow Tuple or NamedTuple `show` to do the rendering of brackets etc
+    show(io, backing(comp))
+end
+
+#TODO think about this, for if we are missing fields
+#Base.convert(::Type{Primal}, comp::Composite{Primal})
+Base.convert(::Type{<:NamedTuple}, comp::Composite{<:Any, <:NamedTuple}) = backing(comp)
+Base.convert(::Type{<:Tuple}, comp::Composite{<:Any, <:Tuple}) = backing(comp)
+
+Base.getindex(comp::Composite, idx) = getindex(backing(comp))
```

Comment:
Should be `backing(comp)[idx]`?

Reply:
```suggestion
Base.getindex(comp::Composite, idx) = getindex(backing(comp), idx)
```

Clearly I need to improve test coverage here.